### PR TITLE
Implementation of transformer model

### PR DESCRIPTION
**Pull request by Bartzi**
This PR adds an example implementation of the Transformer, as introduced in the Paper [Attention is all you need](http://papers.nips.cc/paper/7181-attention-is-all-you-need).

Let me know, if there is anything I should change or add.

### CODE REVIEW

File: transformer/updater/copy_transformer_updater.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import chainer
+import chainer.functions as F
+
+from chainer.training import StandardUpdater
+
+
+class CopyTransformerUpdater(StandardUpdater):
+
+    def update_core(self):
+        with self.device.device:
```
Comment (soskek) [inline transformer/updater/copy_transformer_updater.py]: This part may work on GPU only.
If this script runs on CPU, it causes an error.
```
AttributeError: 'CpuDevice' object has no attribute 'device'
```

Could you solve this kind of errors? Or, recommend users to use GPU. Currently, the default argparse selects CPU (-1).

**Review by soskek — CHANGES_REQUESTED** [review]
Thank you for your PR! It seems good at a glance while it deals with a toy copy task only.

**Comment by Bartzi** [comment]
I changed all occurences of this problem.
Please have a look again =)

**Comment by soskek** [comment]
Thanks!

In a quick check, I found this doesn't seem to work when I set args.gpu != 0 in machines with one or more GPUs. I guessed this may be derived from usages of `with cuda.get_device(self.device):` and/or `with cuda.Device(self.device):`.

### args.gpu = -1
i.e. using CPU while the machine has GPU

```
  File "/home/sosk/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/chainer/training/extensions/evaluator.py", line 161, in __call__
    result = self.evaluate()
  File "/home/sosk/models/transformer/evaluation/copy_transformer_eval_function.py", line 55, in evaluate
    eval_func(data=data, label=data)
  File "/home/sosk/models/transformer/evaluation/copy_transformer_eval_function.py", line 16, in __call__
    with cuda.Device(self.device):
  File "cupy/cuda/device.pyx", line 126, in cupy.cuda.device.Device.__enter__
  File "cupy/cuda/device.pyx", line 141, in cupy.cuda.device.Device.use
  File "cupy/cuda/runtime.pyx", line 193, in cupy.cuda.runtime.setDevice
  File "cupy/cuda/runtime.pyx", line 145, in cupy.cuda.runtime.check_status
cupy.cuda.runtime.CUDARuntimeError: cudaErrorInvalidDevice: invalid device ordinal
```

### args.gpu = 2

i.e. using GPU whose device id is 2 when the machine may have more than 2 GPUs

```
Traceback (most recent call last):
  File "train_copy.py", line 56, in <module>
    main(args)
  File "train_copy.py", line 42, in main
    trainer.run()
  File "/home/sosk/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/chainer/training/trainer.py", line 349, in run
    six.reraise(*exc_info)
  File "/home/sosk/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/home/sosk/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/chainer/training/trainer.py", line 316, in run
    update()
  File "/home/sosk/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/chainer/training/updaters/standard_updater.py", line 175, in update
    self.update_core()
  File "/home/sosk/models/transformer/updater/copy_transformer_updater.py", line 12, in update_core
    self.update_net()
  File "/home/sosk/models/transformer/updater/copy_transformer_updater.py", line 22, in update_net
    predictions = net(batch['data'], batch['label'][:, :-1])
  File "/home/sosk/models/transformer/chains/copy_transformer.py", line 42, in __call__
    result = self.model(x, t, None, self.mask)
  File "/home/sosk/models/transformer/transformer/encoder_decoder.py", line 30, in __call__
    tgt_mask
  File "/home/sosk/models/transformer/transformer/encoder_decoder.py", line 37, in decode
    return self.decoder(self.tgt_embeddings(tgt), memory, src_mask, tgt_mask)
  File "/home/sosk/models/transformer/transformer/decoder.py", line 30, in __call__
    x = layer(x, memory, src_mask, tgt_mask)
...
...
  File "/home/sosk/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/chainer/functions/array/where.py", line 90, in where
    z, = Where(condition).apply((x, y))
  File "/home/sosk/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/chainer/function_node.py", line 321, in apply
    outputs = self.forward(in_data)
  File "/home/sosk/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/chainer/functions/array/where.py", line 36, in forward
    return xp.where(condition, x, y),
  File "/home/sosk/.pyenv/versions/anaconda3-5.0.1/lib/python3.6/site-packages/cupy/sorting/search.py", line 136, in where
    return _where_ufunc(condition.astype('?'), x, y)
  File "cupy/core/core.pyx", line 318, in cupy.core.core.ndarray.astype
  File "cupy/core/core.pyx", line 389, in cupy.core.core.ndarray.astype
  File "cupy/core/_kernel.pyx", line 811, in cupy.core._kernel.ufunc.__call__
  File "cupy/core/_kernel.pyx", line 94, in cupy.core._kernel._preprocess_args
ValueError: Array device must be same as the current device: array device = 0 while current = 2
```



And, if you're involved with PRs, force-push seems not a good choice, because comparing the changes can be difficult (while we can compare with [a github feature](https://github.com/chainer/models/compare/3b58974cddbe282294a59af2598eb5c51d982db7..e4eef8fbda5834dee5c7ae7e14dff92b32fdd6e9)).

**Comment by Bartzi** [comment]
Thanks for reviewing again.
You are totally right, the force push was not the smartest idea :sweat_smile:.
I had a look at the GPU stuff again and I think I finally found a working solution, although I'm not really happy with it. The API seems to be totally messed up in the newer versions of Chainer. Am I using them wrong? There should be a nicer way to write CPU/GPU agnostic code, shouldn't it?

**Comment by soskek** [comment]
Sorry for late reply. I missed it.

Did you see APIs used in any Chainer example code?
https://github.com/chainer/chainer/blob/v6.0.0/examples/mnist/train_mnist.py#L56-L69
It easily selects a device as default in the script. This API seems one of the best ways for simplifying your code.

```
device = chainer.get_device(args.device)
...
model.to_device(device)
device.use()
```

**Comment by Bartzi** [comment]
I did not see this. Seems to be the new way of handling devices. Thanks for the heads up! I will change the implementation as soon as I have time!

**Comment by Bartzi** [comment]
Sorry that it took me so long. I've updated the CPU/GPU usage code and adapted it to the new style, which is used since Chainer v6. Please have a look again.

File: transformer/README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+## Using the Example
+
+If you want to try the example, you'll need to do the following:
+
+0. create a new virtual environment, using Python 3 (preferably >3.6)
+. Python 2 might also work, but this is not guaranteed.
+1. install all requirements via `pip install -r requirements.txt`
+1. start the training with `python train_copy.py`. If you want to use
+a GPU for training, you can specify the GPU with `-g`. For further options
```
Comment (soskek) [inline transformer/README.md]: Now, it is `-d` rather than `-g`.

**Review by soskek — CHANGES_REQUESTED** [review]
Thank you! It works well.
Only one comment.

**Comment by Bartzi** [comment]
Thanks for having a look again and spotting this mistake. I changed the README accordingly.

**Review by soskek — APPROVED** [review]
Thank you for the great work! Merged!

**Pull request merged**